libsignal-service/src/cipher.rs

```diff
@@ -83,84 +97,99 @@ impl ServiceCipher {
         };
 
         use crate::proto::envelope::Type;
-        let (address, metadata) = match envelope.r#type() {
-            Type::PrekeyBundle | Type::Ciphertext => {
-                // is prekey signal message || is signal message
-                (
-                    self.get_preferred_protocol_address(
-                        envelope
-                            .source_address()
-                            .expect("Envelope with source"),
-                        envelope.source_device(),
-                    )?,
-                    Metadata {
-                        sender: envelope
-                            .source_address()
-                            .expect("envelope with source"),
-                        sender_device: envelope.source_device(),
-                        timestamp: envelope.timestamp(),
-                        needs_receipt: false,
-                    },
-                )
-            }
-            Type::UnidentifiedSender => {
-                // is unidentified sender
-                return Err(ServiceError::InvalidFrameError {
-                    reason: "UnidentifiedSender requires SealedSessionCipher, see: https://github.com/Michael-F-Bryan/libsignal-service-rs/issues/10".into(),
-                });
-            }
-            _ => {
-                // else
-                return Err(ServiceError::InvalidFrameError {
-                    reason: "Envelope has unknown type.".into(),
-                });
-            }
-        };
-
-        let mut plaintext = match envelope.r#type() {
+        let plaintext = match envelope.r#type() {
             Type::PrekeyBundle => {
+                let sender = self.get_preferred_protocol_address(
+                    envelope.source_address().expect("Envelope with source"),
+                    envelope.source_device(),
+                )?;
```

Comment:
Duplicate code with `Type::Ciphertext`. Does it make sense to put this before the match? I guess the compiler is smart enough to only call it on those two types.

Reply:
Same with metadata afaict

Reply:
Yeah, I moved it for better readability, I didn't really like the double `match` from before with the second one having a catch all saying `unreachable!("we checked that before")`. Let me see if I come up with something more elegant.

Reply:
You figured out this yet?

Reply:
Not worth the extra effort, as the branch from `UnidentifiedSender` is using a `ProtocolAddress` so you cannot easily have a match arm before to return `(sender, metadata)` without un-necessarily converting types.